unixfs/src/file/adder.rs

```diff
@@ -489,80 +510,138 @@ impl BalancedCollector {
+                    first_at += starting_point.unwrap_or(0);
+
+                    if !finishing && pending[first_at..].len() <= self.branching_factor {
+                        if let Some(last_overwrite) = last_overwrite {
+                            // drain any processed
+                            pending.drain((last_overwrite + 1)..first_at);
+                        }
+                        break 'outer;
+                    }
+
+                    reused_links.clear();
+                    reused_blocksizes.clear();
+
+                    let mut nested_size = 0;
+                    let mut nested_total_size = 0;
+
+                    let last = (first_at + self.branching_factor).min(pending.len());
+
+                    for (index, link) in pending[first_at..last].iter().enumerate() {
+                        assert_eq!(
+                            link.depth,
+                            level,
+                            "unexpected link depth {} when searching at level {} index {}",
+                            link.depth,
+                            level,
+                            index + first_at
+                        );
+
+                        Self::partition_link(
+                            link,
+                            &mut reused_links,
+                            &mut reused_blocksizes,
+                            &mut nested_size,
+                            &mut nested_total_size,
+                        );
+                    }
+
+                    debug_assert_eq!(reused_links.len(), reused_blocksizes.len());
+
+                    let inner = FlatUnixFs {
+                        links: reused_links,
+                        data: UnixFs {
+                            Type: UnixFsType::File,
+                            filesize: Some(nested_size),
+                            blocksizes: reused_blocksizes,
+                            ..Default::default()
+                        },
+                    };
+
+                    let (cid, vec) = render_and_hash(&inner);
+
+                    // start overwriting at the first index of this level, then continue forward on
+                    // next iterations.
+                    let index = last_overwrite.map(|i| i + 1).unwrap_or(first_at);
+                    pending[index] = Link {
+                        depth: level + 1,
+                        target: cid.clone(),
+                        total_size: nested_total_size + vec.len() as u64,
+                        file_size: nested_size,
+                    };
+
+                    ret.push((cid, vec));
+
+                    reused_links = inner.links;
+                    reused_blocksizes = inner.data.blocksizes;
+
+                    starting_point = Some(last);
+                    last_overwrite = Some(index);
+                } else {
+                    // there was nothing to compress on this level.
+                    // FIXME: this is not hit on the tests
```

Comment:
```suggestion
                    // FIXME: this is not hit in the tests
```

Reply:
Now that I changed this from `loop` to `while let Some(...) = ...` I think I see why this shouldn't be hit in the tests... Maybe. Could be that this should have been hit. Resolving either way, I cannot think of anything to hit this.